editor/editor/mode/visual.js

```diff
@@ -1,34 +1,105 @@
 /**
+ * External dependencies
+ */
+import classNames from 'classnames';
+
+/**
  * Internal dependencies
  */
 import InserterButton from '../../inserter/button';
 
-function Blocks( { blocks, onChange } ) {
-	const onChangeBlock = ( index ) => ( changes ) => {
-		const newBlock = {
-			...blocks[ index ],
-			changes
+class Blocks extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.resetBlockState = this.resetBlockState.bind( this );
+		this.state = {
+			blocks: [],
+			selected: null,
+			hovered: null,
+		};
+	}
+
+	componentWillMount() {
+		this.setState( { blocks: this.props.blocks } );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.blocks !== this.state.blocks ) {
+			this.setState( { blocks: nextProps.blocks } );
+		}
+	}
+
+	onChangeBlock() {
+		return ( index ) => ( changes ) => {
+			const { blocks } = this.state;
+			const newBlock = {
+				...blocks[ index ],
+				attributes: {
+					...blocks[ index ].attributes,
+					...changes
+				}
+			};
+
+			const newBlocks = [
+				...blocks.slice( 0, index ),
+				newBlock,
+				...blocks.slice( index + 1 )
+			];
+
+			this.setState( { blocks: newBlocks } );
```

Comment:
It was originally like that But I don't know, I think the `blocks`, `selected`, `hovered` and more to come should stay on the same object (The visual editor's state) which could be updated using a "command" pattern (similar to redux) (See the `updater` in the `per-block` prototypes. 

This allows us to handle undo/redo easily for example. Also having the `selected`, `hovered` state values on the parent component is not a good option because they're not needed in the text mode. 

Thoughts?